ethcore/private-tx/src/private_transactions.rs

```diff
@@ -15,129 +15,182 @@
-	/// 1. Store descriptors inside `VerifiedTransaction`
-	/// 2. Use custom `ready` implementation to only fetch one transaction per sender.
-	/// 3. Get rid of passing dummy `block_number` and `timestamp`
-	transactions: pool::TransactionQueue,
+	verification_pool: RwLock<Pool>,
+	verification_options: pool::verifier::Options,
 }
 
 impl Default for VerificationStore {
 	fn default() -> Self {
 		VerificationStore {
-			descriptors: Default::default(),
-			transactions: pool::TransactionQueue::new(
-				pool::Options {
-					max_count: MAX_QUEUE_LEN,
-					max_per_sender: MAX_QUEUE_LEN / 10,
-					max_mem_usage: 8 * 1024 * 1024,
-				},
-				pool::verifier::Options {
-					// TODO [ToDr] This should probably be based on some real values?
-					minimal_gas_price: 0.into(),
-					block_gas_limit: 8_000_000.into(),
-					tx_gas_limit: U256::max_value(),
-					no_early_reject: false
-				},
-				pool::PrioritizationStrategy::GasPriceOnly,
-			)
+			verification_pool: RwLock::new(
+				txpool::Pool::new(
+					txpool::NoopListener,
+					pool::scoring::NonceAndGasPrice(pool::PrioritizationStrategy::GasPriceOnly),
+					pool::Options {
+						max_count: MAX_QUEUE_LEN,
+						max_per_sender: MAX_QUEUE_LEN / 10,
+						max_mem_usage: 8 * 1024 * 1024,
+					},
+				)
+			),
+			verification_options: pool::verifier::Options {
+				// TODO [ToDr] This should probably be based on some real values?
+				minimal_gas_price: 0.into(),
+				block_gas_limit: 8_000_000.into(),
+				tx_gas_limit: U256::max_value(),
+				no_early_reject: false,
+			},
 		}
 	}
 }
 
 impl VerificationStore {
 	/// Adds private transaction for verification into the store
 	pub fn add_transaction<C: pool::client::Client>(
-		&mut self,
+		&self,
 		transaction: UnverifiedTransaction,
-		contract: Address,
-		validator_account: Address,
-		private_hash: H256,
+		validator_account: Option<Address>,
+		private_transaction: PrivateTransaction,
 		client: C,
 	) -> Result<(), Error> {
-		if self.descriptors.len() > MAX_QUEUE_LEN {
-			bail!(ErrorKind::QueueIsFull);
-		}
 
-		let transaction_hash = transaction.hash();
-		if self.descriptors.get(&transaction_hash).is_some() {
-			bail!(ErrorKind::PrivateTransactionAlreadyImported);
-		}
-
-		let results = self.transactions.import(
-			client,
-			vec![pool::verifier::Transaction::Unverified(transaction)],
-		);
-
-		// Verify that transaction was imported
-		results.into_iter()
-			.next()
-			.expect("One transaction inserted; one result returned; qed")?;
-
-		self.descriptors.insert(transaction_hash, PrivateTransactionDesc {
-			private_hash,
-			contract,
+		let options = self.verification_options.clone();
+		// Use pool's verifying pipeline for original transaction's verification
+		let verifier = pool::verifier::Verifier::new(client, options, Default::default(), None);
+		let verified_tx = verifier.verify_transaction(pool::verifier::Transaction::Unverified(transaction.clone()))?;
```

Comment:
Needless `clone()` call 

You can either take the `transaction (UnverifiedTransaction)` by reference and clone it or take it by value and move it!